controller/tap/apiserver.go

```diff
@@ -87,32 +87,35 @@ func NewAPIServer(
 // ServeHTTP handles all routes for the APIServer.
 func (a *apiServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	a.log.Debugf("ServeHTTP(): %+v", req)
+	err := a.validate(req)
+	if err != nil {
+		a.log.Debug(err)
+		renderJSONError(w, err, http.StatusBadRequest)
+	} else {
+		a.router.ServeHTTP(w, req)
+	}
+}
 
+// validate ensures that the request should be honored returning an error otherwise.
+func (a *apiServer) validate(req *http.Request) error {
 	// if `requestheader-allowed-names` was empty, allow any CN
 	if len(a.allowedNames) > 0 {
-		validCN := ""
-		clientNames := []string{}
 		for _, cn := range a.allowedNames {
 			for _, clientCert := range req.TLS.PeerCertificates {
-				clientNames = append(clientNames, clientCert.Subject.CommonName)
```

Comment:
This would duplicate each common name in the slice for each additional `cn`, so I refactored the listing of the CNs to only be done if and error will returned.